pennylane/templates/embeddings/iqp.py

```diff
@@ -0,0 +1,260 @@
+            res2 = circuit(features=[1., 2., 3.], pattern=pattern2)
+
+            assert np.allclose(res1, res2)
+
+        **Non-consecutive wires**
+
+        In principle, the user can also pass a non-consecutive wire list to the template.
+        For single qubit gates, the i'th feature is applied to the i'th wire index (which may not be the i'th wire).
+        For the entanglers, the product of i'th and j'th features is applied to the wire indices at the i'th and j'th
+        position in ``wires``.
+
+        For example, for ``wires=[2, 0, 1]`` the ``RZ`` block applies the first feature to wire 2,
+        the second feature to wire 0, and the third feature to wire 1.
+
+        Likewise, using the default pattern, the entangler block applies the product of the first and second
+        feature to the wire pair ``[2, 0]``, the product of the second and third feature to ``[2, 1]``, and so
+        forth.
+
+    """
+    #############
+    # Input checks
+
+    wires = check_wires(wires)
+
+    check_no_variable(features, msg="'features' cannot be differentiable")
+
+    expected_shape = (len(wires),)
+    check_shape(
+        features,
+        expected_shape,
+        msg="'features' must be of shape {}; got {}" "".format(expected_shape, get_shape(features)),
+    )
+
+    check_type(
+        n_repeats, [int], msg="'n_repeats' must be an integer; got type {}".format(type(n_repeats))
+    )
+
+    check_type(
+        pattern,
+        [Sequence, type(None)],
+        msg="'pattern' must be None or a list of wire pairs; got {}".format(pattern),
+    )
+
+    if pattern is not None:
+
+        # check type
+        for p in pattern:
+            check_type(
+                p,
+                [list],
+                msg="'pattern' must be None or a list of wire pairs; got {}".format(pattern),
+            )
+
+            for w in p:
+                check_type(
+                    w,
+                    [int],
+                    msg="'pattern' must be None or a list of wire pairs; got {}".format(pattern),
+                )
+
+        # check shape
+        shape = get_shape(pattern)
+        if len(shape) != 2 or shape[1] != 2:
+            raise ValueError("'pattern' must be a list of pairs of wires; got {}".format(pattern))
+
+    #####################
+
+    if pattern is None:
+        # default is an all-to-all pattern
+        pattern = [list(wire_pair) for wire_pair in combinations(wires, 2)]
+
+    for i in range(n_repeats):
+
+        # first block of Hadamards
+        broadcast(unitary=Hadamard, pattern="single", wires=wires)
+        # encode features into block of RZ rotations
+        broadcast(unitary=RZ, pattern="single", wires=wires, parameters=features)
+
+        # entangling block
+        products = []
+        for wire_pair in pattern:
+            # get the position of the wire indices in the array
+            idx1 = wires.index(wire_pair[0])
+            idx2 = wires.index(wire_pair[1])
```

Comment:
It seems like `wires` can be anything classed as a `collections.abc.Iterable` (as checked in `check_wires` in `pennylane.templates.utils`), which thus also would include `np.ndarray`, although the latter don't support `.index` and would cause an error here.

Although the docstring say's that wires should be a `Sequence` which doesn't (?) include `np.ndarrays`, so maybe this is a minor issue.

Reply:
Otherwise I think this solution looks great. :+1:

Reply:
@thisac the user will get an error if using ``np.array`` in the first place, so I hope we can exclude this case?

All of this will hopefully be much easier when we get a new wire management strategy!

Reply:
Ah, you're right. I got confused by the `check_wires` utils function. 

It's apparently the line `if not all([isinstance(w, int) for w in wires])` that passes with `np.ndarray`, since it checks for the `int` type while an `np.ndarray` uses `np.int64`, and thus raises an error.